lib/concurrency/Prefix_Refinement.thy

```diff
@@ -938,163 +1189,614 @@ theorem prefix_refinement_bind_general[rule_format]:
+                            atomized, simplified, rule_format] (*this sequence of attributes removes a trivial assumption*)
+lemmas pfx_refn_bindET =
+  pfx_refn_bindE[where P'="\<top>" and Q'="\<lambda>_ _. True", OF _ _ hoareE_TrueI twp_post_tautE,
+                simplified pred_conj_def, simplified]
+
+lemma prefix_refinement_handle:
+  "\<lbrakk>prefix_refinement sr isr osr (rvr'' \<oplus> rvr) AR R P Q a c;
+    \<And>ft ft'. rvr'' ft ft' \<Longrightarrow> prefix_refinement sr osr osr (rvr' \<oplus> rvr) AR R (E ft) (E' ft') (b ft) (d ft');
+    \<lbrace>P'\<rbrace>,\<lbrace>AR\<rbrace> a -,-,\<lbrace>E\<rbrace>; \<lbrace>Q'\<rbrace>,\<lbrace>R\<rbrace> c -, -,\<lbrace>E'\<rbrace>\<rbrakk>
+   \<Longrightarrow> prefix_refinement sr isr osr (rvr' \<oplus> rvr) AR R (P and P') (Q and Q') (a <handle> (\<lambda>ft. b ft)) (c <handle> (\<lambda>ft'. d ft'))"
+  apply (simp add: handleE_def handleE'_def validIE_def)
+  apply (erule prefix_refinement_bind)
+    defer
+    apply assumption+
+  apply (case_tac v; clarsimp)
+  apply (rule prefix_refinement_return_imp)
+  apply clarsimp
+  done
 
-lemma default_prefix_refinement_ex:
-  "is_matching_fragment sr osr rvr ctr cres s0 R s
-            (\<lambda>s. aprog s \<inter> ({tr'. length tr' \<le> length ctr} \<times> UNIV))
-    \<Longrightarrow> \<exists>f. is_matching_fragment sr osr rvr ctr cres s0 R s f
-            \<and> triv_refinement aprog f"
-  apply (intro exI conjI, assumption)
-  apply (simp add: triv_refinement_def)
+lemma prefix_refinement_catch:
+  "\<lbrakk>prefix_refinement sr isr osr (rvr' \<oplus> rvr) AR R P Q a c;
+    \<And>ft ft'. rvr' ft ft' \<Longrightarrow> prefix_refinement sr osr osr rvr AR R (E ft) (E' ft') (b ft) (d ft');
+    \<lbrace>P'\<rbrace>,\<lbrace>AR\<rbrace> a -,-,\<lbrace>E\<rbrace>; \<lbrace>Q'\<rbrace>,\<lbrace>R\<rbrace> c -, -,\<lbrace>E'\<rbrace>\<rbrakk>
+   \<Longrightarrow> prefix_refinement sr isr osr rvr AR R (P and P') (Q and Q') (a <catch> (\<lambda>ft. b ft)) (c <catch> (\<lambda>ft'. d ft'))"
+  apply (simp add: catch_def validIE_def)
+  apply (erule prefix_refinement_bind)
+    defer
+    apply assumption+
+  apply (case_tac x; clarsimp)
+  apply (rule prefix_refinement_return_imp)
+  apply clarsimp
   done
 
-lemma default_prefix_refinement_ex_match_iosr_R:
-  "is_matching_fragment sr osr rvr ctr cres s0 R s
-            (rely (\<lambda>s. aprog s \<inter> ({tr'. matching_tr_pfx iosr tr' ctr} \<times> UNIV)) R s0)
-    \<Longrightarrow> \<exists>f. is_matching_fragment sr osr rvr ctr cres s0 R s f
-            \<and> triv_refinement aprog f"
-  apply (intro exI conjI, assumption)
-  apply (clarsimp simp add: triv_refinement_def rely_def)
+\<comment> \<open>FIXME: an equivalent doesn't seem to exist for corres, do we want it to or does it mean we
+          shouldn't bother with this?\<close>
+lemma prefix_refinement_handle_elseE:
```

Comment:
I'd keep the rule here. Just because it's not used in the kernel at the moment, doesn't mean we might not need it later, so in terms of library completeness, it should stay. Might be good to add a reminder to add to corrres as well, but not high priority for right now.